shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -484,8 +497,18 @@ - (void)keyUp:(NSEvent*)event {
   [self.flutterViewController keyUp:event];
 }
 
+// Invoked through NSWindow processing of key down event. This can be either
+// regular event sent from NSApplication with CMD modifier, in which case the
+// event is processed as keyDown, or keyboard manager redispatching the event
+// if nextResponder is NSWindow, in which case the event needs to be ignored,
+// otherwise it will cause endless loop.
 - (BOOL)performKeyEquivalent:(NSEvent*)event {
-  return [self.flutterViewController performKeyEquivalent:event];
+  if (_flutterViewController.keyboardManager.eventBeingDispatched == event) {
+    // This happens with cmd+contorl+space (emoji picker)
+    return NO;
+  }
+  [self.flutterViewController keyDown:event];
```

Comment:
What endless loop does it cause? Does it go through `KeyboardManager#handleEvent:`? If so can we detect the loop entirely in `KeyboardManager`?

Reply:
1. `[KeyboardManager dispatchNextEvent:event]` forwards unhandled KeyDown event to `NSWindow`  which is next responder (through (`keyDown:`)
2. `NSWindow` in turn forwards the event to `[FlutterTextInputPlugin performKeyEquivalent:]` (plugin is the first responder).
3. `[FlutterTextInputPlugin performKeyEquivalent:]` forwards the event to [FlutterViewController keyDown:], which then sends it to `KeyboardManager`. That's the loop.

This can't be handled entirely in `KeyboardManager`. The return value of `[FlutterTextInputPlugin performKeyEquivalent:]` is important (that's why it returns NO for redispatched events). For example if the  `TextInputPlugin` returns `YES` for `CMD+CTRL+SPACE` the emoji picker does not show.

Reply:
If it helps here's the stack trace
```
-[FlutterTextInputPlugin performKeyEquivalent:] (/Users/Matej/Projects/flutter/engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:508)
-[NSView performKeyEquivalent:] (@-[NSView performKeyEquivalent:]:54)
-[NSView performKeyEquivalent:] (@-[NSView performKeyEquivalent:]:54)
-[NSWindow _commonPerformKeyEquivalent:conditionally:] (@-[NSWindow _commonPerformKeyEquivalent:conditionally:]:23)
-[NSWindow keyDown:] (@-[NSWindow keyDown:]:12)
-[FlutterKeyboardManager dispatchTextEvent:] (/Users/Matej/Projects/flutter/engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterKeyboardManager.mm:237)
__55-[FlutterKeyboardManager performProcessEvent:onFinish:]_block_invoke (/Users/Matej/Projects/flutter/engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterKeyboardManager.mm:214)
__51-[FlutterChannelKeyResponder handleEvent:callback:]_block_invoke (/Users/Matej/Projects/flutter/engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterChannelKeyResponder.mm:95)
__48-[FlutterBasicMessageChannel sendMessage:reply:]_block_invoke (/Users/Matej/Projects/flutter/engine/src/flutter/shell/platform/darwin/common/framework/Source/FlutterChannels.mm:94)
-[FlutterEngine sendOnChannel:message:binaryReply:]::$_13::operator()(unsigned char const*, unsigned long, void*) const (/Users/Matej/Projects/flutter/engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:841)
```
The event has already been processed by Flutter, we're not interested in it so it is safe (and necessary) to ignore if it gets back to plugin while being redispatched.